### PR TITLE
Use Msgpack Visitor for deserialization.

### LINKED ISSUE #405: Deserialize using less RAM
## Issue with current implementation

Currently, the Deserializer (for both JSON and msgpack deserialization) uses the [`msgpack::unpack`](https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_unpacker) pattern. This pattern first parses the entire serialized data, and then stores it into RAM. The upside is that the implementation is easy and all details are abstracted away. The downside is that that unpacked data may become quite large.

## Proposed solution

The [`msgpack::visitor`](https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_visitor) pattern stores the unpacked data directly into the desired target memory, which saves a lot of memory overhead. This would solve the memory overhead problem.

### Details

The user needs to allocate the memory the Deserializer needs to write the data to himself. To do so, he/she needs to have the dataset information decoded before the data itself is parsed. As a result, 2 passes of the parsing using the [`msgpack::visitor`](https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_visitor) pattern are needed: one for decoding the dataset info, and another for actually decoding the component data.

To reduce the introduced computational overhead, an early return during the dataset info parsing once all dataset info is parsed may be necessary.

### Benchmarking

The fact that 2 passes are necessary (see the details section), means that there is a computational overhead compared to a single pass. Benchmarking after the implementation needs to happen to verify that the computational overhead is not too extreme.

## Useful side effects

While it is not the main purpose of this feature request, the [`msgpack::visitor`](https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_visitor) pattern implementation may also solve #404 , because it would prevent the need for the bug in the unpacker.

### PR DESCRIPTION
**Pull request by TonyXiang8787**
Closes #405 

# Design

* Create multiple visitors.
* Terminate visitor when go over the map/array head. Let the offset move.
* Call different visitors (with moved offset) for parse different data.

# Comparison

Compare the CPU time and memory overhead between this PR and the `main`, to load a 2 GB msgpack PGM data into memory.

## Use current main (1.6.24)

```
Time to load: 23.097098350524902 seconds.
Memory usage of file size: 2.302 GB.
Memory usage of buffers: 10.920 GB.
Total unavoidable memory usage: 13.222 GB.
Total memory usage: 31.815 GB.
Memory overhead: 18.593 GB
```

## Use this PR version

```
Time to load: 12.947288274765015 seconds.
Memory usage of file size: 2.302 GB.
Memory usage of buffers: 10.920 GB.
Total unavoidable memory usage: 13.222 GB.
Total memory usage: 13.287 GB.
Memory overhead: 0.065 GB
```

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 1]
```diff
@@ -200,124 +354,209 @@ class Deserializer {
...
-            MetaComponent const& component = dataset_handler_.dataset().get_component(component_key_);
-            attributes_[component.name] = read_component_attributes(component, kv.val);
+    void parse_predefined_attributes(MetaDataset const& dataset, AttributeByteMeta const& attributes) {
+        root_key_ = "attributes";
+        for (auto const& single_component : attributes) {
+            component_key_ = single_component.first;
+            MetaComponent const* const component = &dataset.get_component(component_key_);
+            std::vector<MetaAttribute const*> attributes_per_component;
+            for (element_number_ = 0; element_number_ != static_cast<Idx>(single_component.second.size());
+                 ++element_number_) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:484]: since we now have `<ranges>`, can you try to see whether this works?

```suggestion
            for (element_number_ : std::views::iota(0, static_cast<Idx>(single_component.second.size())) {
```

of not, please use `boost::counting_iterator`

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 2]
```diff
@@ -200,124 +354,209 @@ class Deserializer {
...
-    std::vector<MetaAttribute const*> read_component_attributes(MetaComponent const& component,
-                                                                msgpack::object const& attribute_list) {
-        if (attribute_list.type != msgpack_array) {
-            throw SerializationError{
-                "Each entry of attribute dictionary should be a list for the corresponding component!\n"};
+    DataByteMeta pre_count_data(bool has_is_batch) {
+        DataByteMeta data_counts{};
+        auto const root_visitor = parse_map_array<true, true, false>();
+        Idx batch_size{};
+        if (has_is_batch && (is_batch_ != !root_visitor.is_map)) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:498]: this feels a bit unreadable to me. why can't you just do `is_batch_ == root_visitor.is_map`? or make a separate function for this

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 3]
```diff
@@ -200,124 +354,209 @@ class Deserializer {
...
+        component_key_ = {};
+        root_key_ = {};
     }
 
-    std::vector<MetaAttribute const*> read_component_attributes(MetaComponent const& component,
-                                                                msgpack::object const& attribute_list) {
-        if (attribute_list.type != msgpack_array) {
-            throw SerializationError{
-                "Each entry of attribute dictionary should be a list for the corresponding component!\n"};
+    DataByteMeta pre_count_data(bool has_is_batch) {
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:494]: this function does a lot of things. can it be split up?

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 4]
```diff
@@ -364,16 +603,16 @@ class Deserializer {
             // So we disable the good code and write the loop manually
             // std::transform_inclusive_scan(msg_data.cbegin(), msg_data.cend(),
             //                               buffer.indptr.begin() + 1,
-            //                               std::plus{}, [](auto const& x) { return static_cast<Idx>(x.size()); });
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:367]: can you try to check whether this still is an issue since we now bumped to newer apple clang? if so, we can remove these comments

File: tests/cpp_unit_tests/test_deserializer.cpp [hunk 5]
```diff
@@ -435,10 +435,12 @@ TEST_CASE("Deserializer with error") {
             R"({"version": "1.0", "type": "input", "is_batch": false, "attributes": {}, "data": {"node": [[5]]}})";
         check_error(unequal_attributes, "Position of error: data/node/0");
         constexpr std::string_view wrong_type_list =
-            R"({"version": "1.0", "type": "input", "is_batch": false, "attributes": {"node": ["id"]}, "data": {"node": [[true]]}})";
+            R"({"version": "1.0", "type": "input", "is_batch": false, "attributes": {"node": ["id"]}, "data": {"node":
+[[true]]}})";
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_deserializer.cpp:439]: why was this change necessary? it's a raw string, so it may not produce the exact same result (newline characters matter). It's good to test this edge case, but I feel like line length is the wrong reason to do that.

**Review by mgovers — COMMENTED** [review]
great improvement. Most of the code looks good, but I would like to see some cleanup still.

File: tests/cpp_unit_tests/test_deserializer.cpp (hunk 5 above)
Comment (TonyXiang8787) [reply tests/cpp_unit_tests/test_deserializer.cpp:439]: The new clang-format automatically formatted this. I guess that's a upgrade of clang-format?

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 6]
```diff
@@ -88,16 +35,216 @@ constexpr from_msgpack_t from_msgpack;
...
+    bool visit_nil() { return throw_error(); }
+    bool visit_boolean(bool /*v*/) { return throw_error(); }
+    bool visit_positive_integer(uint64_t /*v*/) { return throw_error(); }
+    bool visit_negative_integer(int64_t /*v*/) { return throw_error(); }
+    bool visit_float32(float /*v*/) { return throw_error(); }
+    bool visit_float64(double /*v*/) { return throw_error(); }
+    bool visit_str(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool visit_bin(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool visit_ext(const char* /*v*/, uint32_t /*size*/) { return throw_error(); }
+    bool start_array(uint32_t /*num_elements*/) { return throw_error(); }
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:70]: Use name mask is the standard way of visitors. You have some base visitor class with all the (non-virtual) visitor functions defined. If you want to have modified behavior on some of the functions, you write a derived class with definition of those functions. The defined functions in child class will mask and supersede the same function in the base class. For the functions which are not defined in child class, the base class provides the implementation.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 7]
```diff
@@ -88,16 +35,216 @@ constexpr from_msgpack_t from_msgpack;
...
+        std::stringstream ss;
+        ss << msg << ", parsed_offset: " << parsed_offset << ", error_offset: " << error_offset << ".\n";
+        return ss.str();
+    }
+
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    void parse_error(size_t parsed_offset, size_t error_offset) {
+        throw SerializationError{msg_for_parse_error(parsed_offset, error_offset, "Error in parsing")};
+    }
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:52]: Just like below. All the functions are already defined as non-static function in base visitor class. You don't want to write a static function in derived class to mask a non-static function in base class.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:1]: That's a good one. But let's keep it out of the scope of this PR. I created a new issue to expand benchmark with serialization in #421.

[the remark this answers is not in the document]
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:1]: in that case, don't they need to be `override`?

File: tests/cpp_unit_tests/test_deserializer.cpp (hunk 5 above)
Comment (mgovers) [reply tests/cpp_unit_tests/test_deserializer.cpp:439]: interesting

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:70]: in that case, is it correct to derive from `msgpack::null_visitor`? if i remember correctly, the `msgpack` documentation only recommends that if the amount of overloaded functions remains small. I do think that in this implementation, we pretty much override everything, right? In that case, we can take ownership and just make things have `constexpr` implementations instead of hiding or virtual overloads

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:1]: What do you mean `override` here in a benchmark?

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:52]: right, i meant to say it here https://github.com/PowerGridModel/power-grid-model/pull/420#discussion_r1383257150

> in that case, don't they need to be override?

[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:52]: There is no override here. `override` can only be used for virtual function for runtime polymorphism.

We do not use runtime dispatch. Everything is resolved at compile time by name masking.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:70]: Name masking is resolved at compile time. We have defined the `DefaultNullVisitor` which has custom parse error reporting. We inherit from that again to raise error by default for all the visitor functions. We inherit again to mask individual visitor functions for a given type.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 8]
```diff
@@ -200,124 +354,209 @@ class Deserializer {
...
+        std::string_view dataset;
+        Idx batch_size{};
+        Idx global_map_size = parse_map_array<true, false, true>().size;
+        AttributeByteMeta attributes;
+        DataByteMeta data_counts{};
+        // NOLINTNEXTLINE(readability-isolate-declaration)
+        bool has_version{}, has_type{}, has_is_batch{}, has_attributes{}, has_data{};
+
+        while (global_map_size-- != 0) {
+            std::string_view const key = parse_string();
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:405]: It is very hard. When you are traversing the `msgpack` bytes, you do not know which key will come first. If you only want to go through it once, this is the only way I can think of.

[the remark this answers is not in the document]
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:417]: maybe we can implement a map that contains kv pairs of the string and then the action that comes with it, e.g.:

```cpp
auto const actions = std::map{
  {"version", set_version},
  {"is_batch", set_is_batch},
};
```

and then at the end, some function `verify_validity()`

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 2 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:510]: Because you do not know if the key `is_batch` will come first or the key `data`. This line basically means, if `is_batch` comes first, we already have a `is_batch_` value, we need to check if this is consistent with the `data`.

But your comparison is correct, I will change that.

Adding a separate function for this seems to be overkill.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 9]
```diff
@@ -200,124 +354,209 @@ class Deserializer {
...
+            } else {
+                // null visitor to skip
+                parse_skip();
+            }
+            root_key_ = {};
+        }
+
+        if (!has_version) {
+            throw SerializationError{"Key version not found!\n"};
+        }
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:452]: Only `batch_size` and `dataset` are local variables, others are class variables.

I would prefer to keep the style of code here consistent. It is verbose, but quite understandable what is happening.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 4 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:367]: Still fails. After some digging into [cpp reference](https://en.cppreference.com/w/cpp/algorithm/transform_inclusive_scan), we might have done something wrong before.

We did not provide `T init` in the call. It will be inferred by the `decltype(first)::value_type)`, which should not compile. So apple clang is actually giving the correct error.

Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:367]: Let's see after adding `init` if everything can compile.

Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:367]: it compiles!

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 10]
```diff
@@ -139,9 +281,19 @@ class Deserializer {
         } catch (std::exception& e) {
             handle_error(e);
         }
+        root_key_ = {};
     }
 
   private:
+    // data members are order dependent
+    // DO NOT modify the order!
```
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:309]: There are two order dependency here:

`buffer_from_json_` needs to be before `data_` and `size_`. Since we read the `msgpack` buffer twice, the buffer has to live as long as the deserializer lives. If you get a input from `json`, we need to locally store the converted `msgpack` bytes. Only after that, we can save the `data_` and `size_`.

Another dependency is: `dataset_handler_`. This has to be at the end of the members. We are calling many member functions to pre-parse the data to get the information to construct `dataset_handler_`:

```c++
dataset_handler_{pre_parse()}
```

During `pre_parse` other member data variables are used to store meta data and error diagnostic information. Other member data has to be initialized before `pre_parse` is called when the contrcutor tries to initialize `dataset_handler_`.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 3 above)
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:513]: done

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp
[the remark this answers is not in the document]
Comment (TonyXiang8787) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:1]: Many sonar cloud smells are about name hiding. As I explained, name hiding is the standard way for visitors. Even with multiple levels of name hiding.

I have fixed most of other smells.

**Comment by TonyXiang8787** [comment]
@mgovers this PR is ready for review again.

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:309]: member variables should all have a lifetime that is as long as the object right? (or at least until the end of the user-defined body, if any, of the destructor) After the deserializer object is destroyed, we shouldn't need to touch any of the variables, or am i understanding this incorrectly?

**Comment by mgovers** [comment]
Full clang-tidy build started in https://github.com/PowerGridModel/power-grid-model/actions/runs/6795524535

**Review by mgovers — APPROVED** [review]
LGTM. it's still somewhat difficult code at moments but i guess that's to be expected (and probably also the reason almost no libraries expose a visitor pattern deserialization API haha)

clang-tidy also passes https://github.com/PowerGridModel/power-grid-model/actions/runs/6795524535
